### PR TITLE
Do t.storage.Flush() on torrent completion, and on storage.Close() 

### PR DESCRIPTION
**Pull request by AskAlexSharov**
Only question: t.updateComplete is inside lock, is it ok to call syncronious Flush (go-mmap lib doesn't have method for `unix.MS_ASYNC`) there?

### CODE REVIEW

File: mmap_span/mmap_span.go [hunk 1]
```diff
@@ -19,11 +19,30 @@ func (ms *MMapSpan) Append(mMap mmap.MMap) {
...
+	ms.mu.Lock()
+	defer ms.mu.Unlock()
+	for _, mMap := range ms.mMaps {
+		err := mMap.Flush()
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+	// This is for issue 211.
+	ms.mMaps = nil
```
Comment (anacrolix) [inline mmap_span/mmap_span.go:32]: This seems strange, why woudl Flush do this, especially if htere is no error?

File: torrent.go [hunk 2]
```diff
@@ -2461,7 +2461,11 @@ func (t *Torrent) pieceRequestIndexOffset(piece pieceIndex) RequestIndex {
 }
 
 func (t *Torrent) updateComplete() {
+	before := t.Complete.Bool()
```
Comment (anacrolix) [inline torrent.go:2464]: I think there might be a more suitable place to put this, there should be a method that's only called when torrent completion transitions to complete from incomplete.

File: bad_storage.go [hunk 3]
```diff
@@ -17,6 +17,7 @@ var _ storage.ClientImpl = badStorage{}
 func (bs badStorage) OpenTorrent(*metainfo.Info, metainfo.Hash) (storage.TorrentImpl, error) {
 	return storage.TorrentImpl{
 		Piece: bs.Piece,
+		Flush: func() error { return nil },
```
Comment (anacrolix) [inline bad_storage.go:20]: I think it's common enough not to have a Flush method, that teh caller can check for Flush != nil in the TorrentImpl struct.

**Comment by anacrolix** [comment]
I think maybe the asynchronous form should be preferred, unless the client lock is not held. I can take another more careful look when the PR has progressed some more.

File: torrent.go (hunk 2 above)
Comment (AskAlexSharov) [reply torrent.go:2464]: sorry, I don't see such method.

File: bad_storage.go (hunk 3 above)
Comment (AskAlexSharov) [reply bad_storage.go]: done

File: mmap_span/mmap_span.go (hunk 1 above)
Comment (AskAlexSharov) [reply mmap_span/mmap_span.go]: yes, mistake. removed.

File: torrent.go (hunk 2 above)
Comment (anacrolix) [reply torrent.go:2464]: Ah I see the confusion. This flushes the entire torrent, when the entire torrent is complete. Do you want perhaps flush the individual memory-maps, when associated pieces become complete? That would be a more gradual approach.

Comment (anacrolix) [reply torrent.go:2464]: Sorry for not spotting this before. Similarly, if this was the approach, the Flush would move on to the piece implementation. Let me know your thoughts.

**Comment by AskAlexSharov** [comment]
done

**Comment by anacrolix** [comment]
Looks like this PR is turning out to be a bit of a struggle. I took a good look, I can't push to your branch without you letting collaborators edit, but I think I found where you want the flush to occur (for piece completion). See https://github.com/anacrolix/torrent/commit/2c56fa7acb915dd81e0b4903bd13b50c2d7ee479. Let's address that first, and then do the sync on Close later when it is simpler. I notice this PR currently doesn't use the flush at all, that needs to be linked back up via the Piece storage wrapper.

**Comment by AskAlexSharov** [comment]
done

**Comment by AskAlexSharov** [comment]
@anacrolix race between: 
`if p.hasDirtyChunks() {`
and
`piece.unpendChunkIndex(`
I don't see any existing mutex in piece
and probably move `if p.hasDirtyChunks() {` before `t.cl.unlock()` will not help

how better handle it?

**Comment by anacrolix** [comment]
It was just demonstration code for where to put it. You probably do want to check for dirty pieces inside the locked region, then pass out a boolean for it to do the flushing.

**Comment by AskAlexSharov** [comment]
CI green

**Pull request merged**